Eviction admitter: Deny when triggering VMI evacuation

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
When the criteria for VMI evacuation are met (see  https://github.com/kubevirt/kubevirt/pull/11286):
1. The eviction admitter marks the VMI for evacuation.
2. The eviction admitter approves the eviction request.
3. kube-api [denies](https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/#how-api-initiated-eviction-works) the eviction request because the virt-launcher is protected by a PDB.

When a [descheduler](https://github.com/kubernetes-sigs/descheduler) tries to evict a virt-launcher pod owned by a VMI that could be evacuated, it receives a denial for the eviction request and cannot understand that a VMI evacuation was triggered in the background.

After this PR:
When a VMI is marked for evacuation the eviction request is denied with a spacial message.
This is done in order to:
1. Signal the descheduler that an evacuation was triggered although the eviction was denied.
2. Use the existing K8s eviction API

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->
This PR implements the first stage of https://github.com/kubevirt/community/pull/258

### Special notes for your reviewer
~~Depends on PR https://github.com/kubevirt/kubevirt/pull/11380, please skip the first ten commits.~~
Only the last commit changes behavior, other commits are small refactoring.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [X] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [X] PR: The PR description is expressive enough and will help future contributors
- [X] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [X] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

orelmisan commented:
/cc @fossedihelm

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:
```diff
@@ -82,19 +85,16 @@ func (admitter *PodEvictionAdmitter) Admit(ar *admissionv1.AdmissionReview) *adm
 }
 
-func (admitter *PodEvictionAdmitter) markVMI(ar *admissionv1.AdmissionReview, vmiName, nodeName string, dryRun bool) (err error) {
+func (admitter *PodEvictionAdmitter) markVMI(ar *admissionv1.AdmissionReview, vmiName, nodeName string) (err error) {
```
IMHO here we should propagate the dryRun to the `PatchOptions` instead of getting rid of it.
As now, we are not performing the "full" dry run execution.

orelmisan replied:
> IMHO here we should propagate the dryRun to the `PatchOptions` instead of getting rid of it. As now, we are not performing the "full" dry run execution.

What will be the added value of doing so?

fossedihelm replied:
The DryRun execution should perform and test all the things that will be performed by a normal execution, without storing anything, so without persisting changes.
It is used to test something.
Right now it is possible that the dryRun execution succeed, so a user thinks that everything will be good, but then the request will fail without dryRun execution due to the Patch request in the `markVMI` function.

orelmisan replied:
I will revert the dryRun to the way it was.
A follow-up PR could perform this change which is unrelated to this PR.

orelmisan replied:
Done.

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:
```diff
@@ -74,7 +74,7 @@ func (admitter *PodEvictionAdmitter) Admit(ar *admissionv1.AdmissionReview) *adm
 		vmi.Status.NodeName == launcher.Spec.NodeName &&
 		!dryRun {
-		if err := admitter.markVMI(ar, vmi.Name, vmi.Status.NodeName); err != nil {
+		if err := admitter.markVMI(vmi.Namespace, vmi.Name, vmi.Status.NodeName); err != nil {
```
this refactoring should be part of a unique refactoring commit of the `markVMI` function.

orelmisan replied:
Done.

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:
```diff
@@ -85,9 +85,9 @@ func (admitter *PodEvictionAdmitter) Admit(ar *admissionv1.AdmissionReview) *adm
 }
 
-func (admitter *PodEvictionAdmitter) markVMI(vmiNamespace, vmiName, nodeName string) (err error) {
+func (admitter *PodEvictionAdmitter) markVMI(vmiNamespace, vmiName, nodeName string) error {
```
same as above

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:
```diff
@@ -36,12 +36,12 @@ func (admitter *PodEvictionAdmitter) Admit(ar *admissionv1.AdmissionReview) *adm
 	launcher := pod
 
-	domainName, exists := launcher.GetAnnotations()[virtv1.DomainAnnotation]
+	vmiName, exists := launcher.GetAnnotations()[virtv1.DomainAnnotation]
```
This commit can belong to a bigger commit with the previous one (refactoring pod eviction admitted `Admit` function)

orelmisan replied:
Squashed it with the first commit.

fossedihelm reviewed (commented):
Hey @orelmisan! While it is great to refactor the code, I think this one was not in the right direction. Also, I think we can squash most of the commits since right now the granularity is too high.
I'd expect here something like:
- refactoring `Admin`
- refactoring `markVMI`
- the "feature"(last) commit.
More details below inline.
Thank you

orelmisan commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:
```diff
@@ -24,15 +25,17 @@ type PodEvictionAdmitter struct {
 		return validating_webhooks.NewPassingAdmissionResponse()
 	}
 
+	launcher := pod
```
> I don't get the real added value here honestly and the launcher := pod is something that I will not encourage.

Currently, it is not possible to filter eviction requests by pod label [1][2], so unfortunately this webhook intercepts all eviction requests in the cluster - including for pods that are not virt-launchers.

Since the eviction request does not contain any data on the evicted pod other than its namespace/name, the webhook has to fetch the pod in order to understand whether its a virt-launcher or not (by reading its `kubevirt.io` label and expecting the value to be `virt-launcher`).

Only after this test, the pod could be considered a virt-launcher.

The current code called the fetched pod `launcher`, even if it is not a virt-launcher pod.
I found it to be confusing and misleading.

> The good refactoring that I'd expect here is to use the labelselector while retrieving the virt-launcher pod

AFAIK, using labelSelectors is useful when performing List operations.
In this case we perform a Get operation to get a single pod.
Could you please elaborate what do you mean in this context?

[1] https://github.com/kubernetes/kubernetes/issues/110169#issuecomment-1140512056
[2] https://kubernetes.slack.com/archives/C0EG7JC6T/p1707054818877809

orelmisan replied:
For completeness, the AdmissionReview object sent from kube-api to the webhook looks like this:
```json
{
  "kind": "AdmissionReview",
  "apiVersion": "admission.k8s.io/v1",
  "request": {
    "uid": "f80b8729-034e-4198-bcc9-1e7aca659ca1",
    "kind": {
      "group": "policy",
      "version": "v1",
      "kind": "Eviction"
    },
    "resource": {
      "group": "",
      "version": "v1",
      "resource": "pods"
    },
    "subResource": "eviction",
    "requestKind": {
      "group": "policy",
      "version": "v1",
      "kind": "Eviction"
    },
    "requestResource": {
      "group": "",
      "version": "v1",
      "resource": "pods"
    },
    "requestSubResource": "eviction",
    "name": "mypod",
    "namespace": "default",
    "operation": "CREATE",
    "userInfo": {
      "username": "kubernetes-admin",
      "groups": [
        "kubeadm:cluster-admins",
        "system:authenticated"
      ]
    },
    "object": {
      "kind": "Eviction",
      "apiVersion": "policy/v1",
      "metadata": {
        "name": "mypod",
        "namespace": "default",
        "creationTimestamp": null,
        "managedFields": [
          {
            "manager": "kubectl",
            "operation": "Update",
            "apiVersion": "policy/v1",
            "time": "2024-02-28T09:32:08Z",
            "fieldsType": "FieldsV1",
            "fieldsV1": {
              "f:deleteOptions": {}
            },
            "subresource": "eviction"
          }
        ]
      },
      "deleteOptions": {}
    },
    "oldObject": null,
    "dryRun": false,
    "options": {
      "kind": "CreateOptions",
      "apiVersion": "meta.k8s.io/v1"
    }
  }
}
```

fossedihelm replied:
Sorry, maybe I misexplained what was not clear to me. The fact that we were excluding non-virtlauncher pods was clear.
I didn't understand what was the reasoning behind this commit.
I think that if a refactoring can be done here is to filter the non-virtlauncher pods serverside, with labelselector.
So something like:
```
	pods, err := admitter.VirtClient.CoreV1().Pods(ar.Request.Namespace).List(context.Background(), metav1.ListOptions{LabelSelector: fmt.Sprintf(virtv1.AppLabel + "=virt-launcher"), FieldSelector: "metadata.name=" + ar.Request.Name})
	if err != nil || pods.Size() == 0 {
		return validating_webhooks.NewPassingAdmissionResponse()
	}

	launcher := pods.Items[0]
```

WDYT?

orelmisan replied:
An eviction request is sent from its initiator, lets say kubectl, per pod [1].
Hence in this webhook, we get the AdmissionReview object derived from a single eviction request, containing only the evicted pod's namespace/name.

This is the reason a Get operation is performed - it gets a single pod by namespace/name.

IMO the List operation is:
1. More costly on kube-api's side.
2. A bit more costly on the network since more JSON is sent.
3. Is less readable.

I don't a clear advantage over the existing approach (which was unchanged by this PR).

[1] https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/#calling-the-eviction-api

fossedihelm replied:
After an offline conversation maybe here is better to continue using `Get` (thanks for the discussion).
But please, remove `launcher := pod` :)

orelmisan replied:
> After an offline conversation maybe here is better to continue using `Get` (thanks for the discussion). But please, remove `launcher := pod` :)

Could you please explain the reason behind your objection? I haven't found it in this thread.
Please note that this copy has very low cost - only a pointer is copied.

orelmisan replied:
```go
if !isVirtLauncher(launcher) {
}
```

fossedihelm replied:
You can always rename the returned value from  the `Get` operation, so that it will be:
```
if !isVirtLauncher(pod) {
}
```

orelmisan replied:
Done.

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:
```diff
@@ -36,7 +36,7 @@ func (admitter *PodEvictionAdmitter) Admit(ar *admissionv1.AdmissionReview) *adm
 	launcher := pod
 
-	vmiName, exists := launcher.GetAnnotations()[virtv1.DomainAnnotation]
+	vmiName, exists := launcher.Annotations[virtv1.DomainAnnotation]
```
To me, generally speaking, using the interface is better then direct access

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/16334d1923ec6b35ae2b17411ae3c863e73884c1..d43fd3e1dabad25673db2e702fc7a2f42d219cf8): Dropped the commit to change `pod.GetAnnotations to pod.Annotations`
Dropped the commit to move the dryRun decision to the calling funciton.
Squashed the commits to change the signature of `markVMI`.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/d43fd3e1dabad25673db2e702fc7a2f42d219cf8..2531fd61d8dc7bc124c6043cc358f6197999863e): Rebase

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/2531fd61d8dc7bc124c6043cc358f6197999863e..4ea48888e6eb8eb3e4701cbc001a6560a7d63cb8): Rebase

orelmisan commented:
@fossedihelm @xpivarc PTAL

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/4ea48888e6eb8eb3e4701cbc001a6560a7d63cb8..676f19a6a0ade26ae8e2b016862bfb6227a400ce): Simplified isVirtLauncher.
Added more details in the first commit message.

fossedihelm reviewed (commented):
Sorry for taking so long for the review. :)

fossedihelm commented:
Thank you!
/lgtm

orelmisan commented:
@xpivarc could you please have a look?

orelmisan commented:
@enp0s3 this is the output when an evacuation is triggered following a node drain (the VMI eviction policy is `LiveMigrate`):
```bash
$ kubectl drain node01 --ignore-daemonsets --delete-emptydir-data
...
evicting pod default/virt-launcher-vm-cirros-wn5v4
error when evicting pods/"virt-launcher-vm-cirros-wn5v4" -n "default" (will retry after 5s): admission webhook "virt-launcher-eviction-interceptor.kubevirt.io" denied the request: Triggered evacuation of VMI "default/vm-cirros"
...
evicting pod default/virt-launcher-vm-cirros-wn5v4
error when evicting pods/"virt-launcher-vm-cirros-wn5v4" -n "default" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
...
evicting pod default/virt-launcher-vm-cirros-wn5v4
pod/virt-launcher-vm-cirros-wn5v4 evicted
node/node01 drained
```

On the first eviction request the VMI is marked for evacuation and the eviction request is denied.
On the second eviction request, the VMI was already marked, so the request is approved by the webhook, and fails because a PDB is in place to protect the source (and target) virt-launcher pod(s).

ingvagabund commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:
```diff
@@ -75,6 +75,10 @@ func (admitter *PodEvictionAdmitter) Admit(ar *admissionv1.AdmissionReview) *adm
 		}
+
+		if !dryRun {
+			return denied(fmt.Sprintf("Triggered evacuation of VMI \"%s/%s\"", vmi.Namespace, vmi.Name))
```
- "Triggered eviction: evacuation of VMI ..."
- "Triggering eviction: evacuation of VMI ..."
- "Eviction triggered: evacuation of VMI ..."
- or similar

either if these will work better. A generic prefix will be easier to parse than a more explicit message.

orelmisan replied:
```suggestion
			return denied(fmt.Sprintf("Eviction triggered evacuation of VMI \"%s/%s\"", vmi.Namespace, vmi.Name))
```

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/bf3fb0ec45e7317e8ba3f44990120037911a51b7..a5d03e9cab3b954c7a2d06de3d6171371061f303):
Changed error message to `"Eviction triggered evacuation of VMI \"%s/%s\""`
Renamed `launcher` -> `pod`

orelmisan commented:
@fossedihelm @ingvagabund PTAL

fossedihelm reviewed (approved):
Thanks

orelmisan commented:
/test pull-kubevirt-unit-test

ingvagabund commented:
Looks good, thx

acardace commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:
```diff
@@ -65,27 +66,32 @@ func (admitter *PodEvictionAdmitter) Admit(ar *admissionv1.AdmissionReview) *adm
+		}
 	}
 
 	// We can let the request go through because the pod is protected by a PDB if the VMI wants to be live-migrated on
```
Can you remove this comment? You now deny the request so the comment does not reflect reality anymore.

orelmisan replied:
This comment is still correct.
For example:
1. This is not the first time the virt-launcher pod is evicted (this happens in a loop when running `kubectl drain`)
2. A virt-launcher pod of a non migratable VMI is evicted.

orelmisan replied:
The full table is located here: https://github.com/kubevirt/kubevirt/blob/main/docs/handling-eviction-requests.md#eviction-approval-summary

kubectl output can be found [here](https://github.com/kubevirt/kubevirt/pull/11532#issuecomment-2094766310).

acardace replied:
Gotcha, I was missing the loop part.

acardace commented on pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go:
```diff
@@ -65,27 +66,32 @@ func (admitter *PodEvictionAdmitter) Admit(ar *admissionv1.AdmissionReview) *adm
 			return denied(fmt.Sprintf("kubevirt failed marking the vmi for eviction: %s", err.Error()))
 		}
+
+		if !dryRun {
```
This is not right, you should deny the request whether it's a dry run request or not.

DryRun mode allows you to see what would happen without making any persistent modification to the system, if you behave differently here it'd be a violation of how a dryrun request should behave.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/a5d03e9cab3b954c7a2d06de3d6171371061f303..e65b39ce2f4f619bc04d53155ca00c56e50b7fc3): Answered @acardace's [comment](https://github.com/kubevirt/kubevirt/pull/11532#discussion_r1595371879) and @fossedihelm's [comment](https://github.com/kubevirt/kubevirt/pull/11532#discussion_r1528736440).

fossedihelm reviewed (approved):
Thank you!

acardace commented:
/approve
